storage-access.bs

```diff
@@ -129,13 +157,17 @@ When invoked on {{Document}} |doc|, the <dfn export method for=Document><code>ha
 <!-- https://hg.mozilla.org/mozilla-central/file/tip/dom/base/Document.cpp#l15512 -->
 
 1. Let |p| be [=a new promise=].
-1. Let |flag set| be the result of [=obtain the storage access flag set|obtaining the storage access flag set=] of |doc|.
-1. If |flag set|'s [=was expressly denied storage access flag=] is set, [=resolve=] |p| with false and return |p|. <!-- WebKit's DocumentStorageAccess.cpp#L85 -->
-1. If |doc|'s [=Document/origin=] is an [=opaque origin=], [=resolve=] |p| with false and return |p|. <!-- WebKit's DocumentStorageAccess.cpp#L90 --> <!-- Gecko's Document.cpp#l15526 -->
-1. If |doc|'s [=Document/browsing context=] is a [=top-level browsing context=], [=resolve=] |p| with true and return |p|. <!-- WebKit's DocumentStorageAccess.cpp#L95 --> <!-- Gecko's Document.cpp#l15531 -->
+1. Let |map| be the result of [=obtain the storage access map|obtaining the storage access map=] for |doc|.
+1. Let |key| be the result of [=generate a partitioned storage key|generating a partitioned storage key=] from |doc|.
+1. If |key| is failure, then [=/resolve=] |p| with false and return |p|.
```

Comment:
Is there a risk of returning the wrong result by checking for an expressly denied flag value here? Since the flag is checked in `[=determine if a site has storage access=]` can we skip these two lines here?

If the result of `[=determine if a site has storage access=]` can vary by implementation, by user setting to block third party storage or other mechanisms, would it be possible to enter here once while the impl logic and lack of "grant" cause hasStorageAccess to return false. As a result requestStorageAccess is called which is then expressly denied. Then the result of the impl logic to grant storage access changes and future calls to hasStorageAccess would still return false despite access actually being present and no grant being required.

Reply:
I think I see what you mean. The [=was expressly denied storage access flag=] needs to get reset somewhere.

Reply:
Does 8cde662 address your concern, @Brandr0id?

Reply:
> Does 8cde662 address your concern

That actually looks like a similar but different instance of the same class of bug :).

In this case I think the nuance is in the implementer defined logic inside of `[=determine if a site has storage access=]`
> 1. Let |has storage access| (a [=boolean=]) be the result of running an [=implementation-defined=] set of steps to determine if |key|'s [=partitioned storage key/embedded site=] has access to its [=first party data=] on |key|'s [=partitioned storage key/top-level site=].

A nice simple example would be the block all third party cookies user option/preference. If the |implementation-defined| logic can vary during the lifetime of the page we run into this issue. I would say that a user and UX should probably prompt/lead to the refresh of already opened tabs but in practice that may not occur. The implementation logic could also just as easily by something that is by nature dynamic like tracking prevention features or other user experiences.

Using the third party cookie option we could have:

1. User has "block all third party cookies" set.
1. Visits a page with embedded 3p content that has storage access blocked so uses document.requestStorageAccess()
    1. Access is blocked, no implicit grant is given due to impl logic
    1. A prompt is shown and rejected leading to explicit denial and a flag being set to prevent future use of requestStorageAccess() on this document.
1. Subsequent calls to document.hasStorageAccess() are correct and return **false**
1. User preference for blocking third party cookies is unset, no reload occurs.
1. Subsequent calls to document.hasStorageAccess() are incorrect as flag is set but implementation logic no longer blocks storage access so **false** is returned but access to storage past this point would work.

Reply:
@Brandr0id thanks, it's clear now, and I've tried to address this in 6c1e4b4.

Reply:
The change adds good clarity to the spec but I don't think it addresses this issue. The issue is the potentially inaccurate response of hasStorageAccess and there is no guarantee a new requestStorageAccess call has been made that might update the flags/map.

Specifically I think I would propose just removing steps 2-6 inclusively (note step 3 appears to be a duplicate of step 11):

![image](https://user-images.githubusercontent.com/18271690/81619388-53e73a80-939e-11ea-9f08-ca8d64a55730.png)

Since when we get execute step 12, we'll take the flag value into consideration along with any implementation specific logic to determine if access is present or not:
![image](https://user-images.githubusercontent.com/18271690/81619459-81cc7f00-939e-11ea-9f86-66657f0d83d6.png)

Actually looking closer here too, I'm not sure if caching the "has storage access flag" or referencing it here doesn't cause potential issues either. 

Since the hasStorageAccess method is trying to reflect the current state of storage access and ultimately gets this from an implementation-defined set of steps I think we may need to always check the implementation-defined steps. Perhaps any early short circuit in the method  that avoids the implementation-defined logic either has an implicit guarantee that the implementation-defined steps don't change throughout the lifetime of the document or risks the cached flag/response being inaccurate.